EIPS/eip-6357.md

```diff
@@ -0,0 +1,79 @@
+---
+eip: 6357
+title: Single-contract Multi-delegatecall
+description: Allows an EOA to call multiple functions of a smart contract in a single transaction
+author: Pandapip1 (@Pandapip1)
+discussions-to: https://ethereum-magicians.org/t/eip-6357-single-contract-multicall/12621
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-01-18
+---
+
+## Abstract
+
+This EIP standardizes an interface containing a single function, `multicall`, allowing EOAs to call multiple functions of a smart contract in a single transaction, and revert all calls if any call fails. 
+
+## Motivation
+
+Currently, in order to transfer several NFTs or SFTs, one needs to submit a number of transactions equal to the number of NFTs being tranferred. This wastes users' funds by requiring them to pay an [EIP-1559](./eip-1559.md) priority fee for every NFT they transfer.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Contracts implementing this EIP must implement the following interface:
+  
+```solidity
+pragma solidity ^0.8.0;
+
+interface IMulticall {
+    /// @notice           Takes an array of abi-encoded call data, delegatecalls itself with each calldata, and returns the abi-encoded result
+    /// @dev              Reverts if any delegatecall reverts
+    /// @param    data    The abi-encoded data
+    /// @returns  results The abi-encoded return values
+    function multicall(bytes[] calldata data) external virtual returns (bytes[] memory results);
+
+    /// @notice           OPTIONAL. Takes an array of abi-encoded call data, delegatecalls itself with each calldata, and returns the abi-encoded result
+    /// @dev              Reverts if any delegatecall reverts
+    /// @param    data    The abi-encoded data
+    /// @param    values  The effective msg.values. These must add up to at most msg.value
+    /// @returns  results The abi-encoded return values
+    function multicallPayable(bytes[] calldata data, uint256[] values) external payable virtual returns (bytes[] memory results);
+}
+```
+
+## Rationale
+
+Needs discussion.
+
+## Backwards Compatibility
+
+This is compatible with most existing multicall functions.
+
+## Reference Implementation
+
+```solidity
+pragma solidity ^0.8.0;
+
+/// Derived from OpenZeppelin's implementation
```

Comment:
How does this affect the license of the code?

Reply:
Not sure; don't really care. OpenZeppelin stole it from some unreleased code in ENS, so I'm sure it's fine.